src/stan/services/util/initialize.hpp

```diff
@@ -13,202 +13,211 @@
+                  " at the initial value.");
+      logger.info(e.what());
+      throw;
+    }
 
-            std::stringstream msg2;
-            msg2 << "1000 transitions using 10 leapfrog steps"
-                 << " per transition would take"
-                 << " " << 1e4 * deltaT << " seconds.";
-            logger.info(msg2);
+    msg.str("");
+    double log_prob(0);
+    try {
+      log_prob = model.template log_prob<false, true>
+                 (unconstrained, disc_vector, &msg);
+      if (msg.str().length() > 0)
+        logger.info(msg);
+    } catch (std::domain_error& e) {
+      if (msg.str().length() > 0)
+        logger.info(msg);
+      logger.info("Rejecting initial value:");
+      logger.info("  Error evaluating the log probability"
+                  " at the initial value.");
+      logger.info(e.what());
+      continue;
+    } catch (std::exception& e) {
+      if (msg.str().length() > 0)
+        logger.info(msg);
+      logger.info("Unrecoverable error evaluating the log probability"
+                  " at the initial value.");
+      logger.info(e.what());
+      throw;
+    }
+    if (!boost::math::isfinite(log_prob)) {
+      logger.info("Rejecting initial value:");
+      logger.info("  Log probability evaluates to log(0),"
+                  " i.e. negative infinity.");
+      logger.info("  Stan can't start sampling from this"
+                  " initial value.");
+      continue;
+    }
+    std::stringstream log_prob_msg;
+    std::vector<double> gradient;
+    bool gradient_ok = true;
+    clock_t start_check = clock();
+    try {
+      log_prob = stan::model::log_prob_grad<true, true>
+                 (model, unconstrained, disc_vector,
+                  gradient, &log_prob_msg);
+    } catch (const std::exception& e) {
+      if (log_prob_msg.str().length() > 0)
+        logger.info(log_prob_msg);
+      logger.info(e.what());
+      throw;
+    }
+    clock_t end_check = clock();
+    double deltaT = static_cast<double>(end_check - start_check)
+                    / CLOCKS_PER_SEC;
+    if (log_prob_msg.str().length() > 0)
+      logger.info(log_prob_msg);
 
-            logger.info("Adjust your expectations accordingly!");
-            logger.info("");
-            logger.info("");
-          }
-          if (gradient_ok)
-            break;
-        }
+    for (size_t i = 0; i < gradient.size(); ++i) {
+      if (gradient_ok && !boost::math::isfinite(gradient[i])) {
+        logger.info("Rejecting initial value:");
+        logger.info("  Gradient evaluated at the initial value"
+                    " is not finite.");
+        logger.info("  Stan can't start sampling from this"
+                    " initial value.");
+        gradient_ok = false;
+      }
+    }
+    if (gradient_ok && print_timing) {
+      logger.info("");
+      std::stringstream msg1;
+      msg1 << "Gradient evaluation took " << deltaT << " seconds";
+      logger.info(msg1);
```

Comment:
[future request]

Our loggers should just take these, as in:

```
logger.info() << "Gradient evaluation took " << ...;
```

Or we should have a polyadic info function that could look like:

```
logger.info("Gradient evaluation took ", deltaT, " seconds");
```

Reply:
I just created a new issue for this.

The first implementation was a compromise and was only meant to be there as a stepping stone. Now that it's in, it's a lot easier to change safely.

Reply:
Yes, we should. The current design was a compromise and a stepping stone. 

I'd like move forward now that this is in.